feature(config): Add go-sockaddr templating support to nomad consul address

Nkmol opened the pull request:
This PR adds support for go-sockaddr/template in the [consul address Stanza](https://www.nomadproject.io/docs/configuration/consul#address).

As noted in the issue, this becomes especially useful when you want to pair your nomad agent to the consul private network agent in a dynamic fashion.

Fixes: #11062


### Special note to the reviewer
A check has been added to only support one single IP in the Stanza, as was done in other places using `ParseSingleIPTemplate`. However, I think in the years Hashicorp defines this in their shared package: https://github.com/hashicorp/go-secure-stdlib/blob/7166b1cdf6a6eb1311247e34feae17eaf7b657bf/listenerutil/parse.go#L374
Adding an extra dependency to the repo of [go-secure-stdlib](https://github.com/hashicorp/go-secure-stdlib) might not be something preferable. If this is still preferable, I could add this extra import.

---

This is my first contribution to this project and the first time touching Go, so any tips or directions for improvement is always welcome :)

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=12084)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=12084)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**sander-mol** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad?pullRequest=12084).</sub>

DerekStrickland commented:
Hi @Nkmol . Thanks for contributing! 

I'm going to take a look today and talk to the team about your contribution.

DerekStrickland reviewed (dismissed):
I found one thing right away that is a blocker. See comment about panic.

Nkmol commented on nomad/structs/config/consul.go:
```diff
@@ -251,7 +253,11 @@ func (c *ConsulConfig) ApiConfig() (*consul.Config, error) {
-		config.Address = c.Addr
+		out, err := template.Parse(c.Addr)
+		if err != nil {
+			panic(fmt.Errorf("unable to parse address template %q: %v", c.Addr, err))
```
Woops, thank you for noticing this 👍

Nkmol commented on helper/config/config.go:
```diff
@@ -0,0 +1,28 @@
+
+// ParseSingleIPTemplate is used as a helper function to parse out a single IP
+// address from a config parameter.
+func ParseSingleIPTemplate(ipTmpl string) (string, error) {
```
Copied from https://github.com/hashicorp/go-secure-stdlib/blob/7166b1cdf6a6eb1311247e34feae17eaf7b657bf/listenerutil/parse.go#L374

shoenig replied:
Instead of moving the copy of the library, I think we could just use the library? The upstream repo scopes each package to it's own module, so we wouldn't pull in much new.

tgross replied:
Hm... I'm kind of very-soft-against because the package pulls in stuff like their tlsutils package, etc. But it looks like at least they're pinned to the same version of go-sockaddr, so maybe that's ok?

tgross replied:
I've slept on it and I'm sold... let's pull in `go-secure-stdlib/listenrutil` here

tgross commented on helper/config/config.go:
```diff
@@ -0,0 +1,28 @@
+package config
```
As long as we're making a new package, maybe it would be nice if this were something we could use for other IP-address-related utilities? What if we called it `helper/ipaddress` instead?

tgross reviewed (changes requested):
Hi @Nkmol! Thanks for this PR, this is looking really good. I've left a few comments but once those are resolved I imagine we can get this shipped as part of Nomad 1.3.0

If you'd like, you can add a changelog file too. That'd be a text file at `.changelog/12084.txt` formatted similar to [.changelog/12079.txt](https://github.com/hashicorp/nomad/blob/main/.changelog/12079.txt).

shoenig commented on nomad/structs/config/consul_test.go:
```diff
@@ -167,3 +169,42 @@ func TestConsulConfig_Exec(t *testing.T) {
+	privateIps, err := sockaddr.GetPrivateIP()
+	if err != nil {
+		t.Fatalf("failed to get private IPs: %v", err)
+	}
```
a lot of old code does this, but in fresh code we've been preferring testify; 

```
require.NoError(t, ...)
```

Nkmol commented:
Thanks @shoenig and @tgross for the thorough review. Based on the feedback, I changed the following:

- [x] **Imported the https://github.com/hashicorp/go-secure-stdlib/releases/tag/listenerutil%2Fv0.1.4 module**
Importing this module, however, introduced more dependencies than I was expecting. My gut feeling is telling me the dependencies added and changed might be too much out of scope of this PR. I am having a had time assessing the `mod.go` changes :) Let me know what you think about this. 
- [x] **Implemented table-driven tests**
TIL :) I did not know about [this](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) and tried to make them as parallel runnable as possible being aware of the [Go Routine bug](https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721).
- [x] **Migrated from Go asserts to [testify](https://github.com/stretchr/testify)**

tgross reviewed (approved):
LGTM! I'll get this merged once CI is done and it'll ship in Nomad 1.3.0. Thanks again @Nkmol!
